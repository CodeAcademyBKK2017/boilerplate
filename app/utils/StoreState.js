const getSelector = (key) => (storeState) => storeState[key];

export const getNotesSelector = getSelector('notes');

export default getSelector;
