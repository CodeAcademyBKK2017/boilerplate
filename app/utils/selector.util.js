const getSelector = (key) => (store) => (store[key]);

export const selectNotes = getSelector('notes'); 

export default getSelector; 