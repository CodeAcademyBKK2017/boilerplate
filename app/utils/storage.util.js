import {AsyncStorage} from 'react-native';
// Keep for references
// const getItem = (key) => AsyncStorage.getItem(key).then(JSON.parse);

const getItem = (key) => AsyncStorage.getItem(key).then((data) => JSON.parse(data));

const setItem = (key, data) => AsyncStorage.setItem(key, JSON.stringify(data));

export default {
  getItem, 
  setItem
};