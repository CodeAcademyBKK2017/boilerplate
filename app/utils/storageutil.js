import {AsyncStorage} from 'react-native';

class storageutil {
  getItem = async () => JSON.parse(await AsyncStorage.getItem('notes')) || []

  setItem = async (newNotes) => {
    await AsyncStorage.setItem('notes', JSON.stringify(newNotes));
  }
} export default new storageutil();