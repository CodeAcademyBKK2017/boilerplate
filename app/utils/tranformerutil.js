class Tranformerutil {
  removeNote (afterDel, id) {
    return afterDel.filter((note) => note.id !== id);
  }
} export default new Tranformerutil();