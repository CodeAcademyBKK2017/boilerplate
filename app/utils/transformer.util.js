export const removeNote = (notes, id) => notes.filter((note) => note.id !== id);

export default {
  removeNote
};