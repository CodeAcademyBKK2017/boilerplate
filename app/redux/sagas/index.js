import ApiNotes from '../../api';
import {Alert} from 'react-native';
import {call, fork, put,  select, take, takeLatest} from 'redux-saga/effects';
import {delay} from 'redux-saga';
import {filterNote} from '../../utils/transformerutil';
import {getItemToStorage, setItemToStorage} from '../../utils/storageutil';
import * as actions from '../actions/index.actions';

function* fetchHandler () {
  yield put({
    type: actions.SHOW_LOADER
  });
  //  yield delay(1000);
  let response;
  try {
    response = yield call(ApiNotes.getNotes);
  } catch (error) {
    const storageData =  yield call(getItemToStorage, 'storageNote');
    response = storageData ? storageData : [];
  }
  yield put({
    type: actions.POPULATE_NOTES,
    payload: response
  });
  yield put({
    type: actions.HIDE_LOADER
  });
}
function* addNoteHandler (action) {
  yield put({
    type: actions.SHOW_LOADER
  });
  yield delay(1000);
  try {
    const note = action.payload;
    const noteWithId = yield call(ApiNotes.addNote, note);
    const currentNote = yield select((store) => (store.notes));
    const newNote = [...currentNote, noteWithId];
    yield put(actions.addNote(noteWithId)); 
    yield call(setItemToStorage, 'storageNote', newNote);
  } catch (error) {
    Alert.alert(
      'Save Failed', 
      String(error),
      [
        {text: 'OK'}
      ],
      {
        cancelable: false
      }
    );
  }
  yield put({
    type: actions.HIDE_LOADER
  });
}
function* deleteNoteHandler (action) {
  yield put({
    type: actions.SHOW_LOADER
  });
  try {
    const itemId = action.payload;
    yield call(ApiNotes.deleteNote, itemId);
    yield put(actions.deleteNote(itemId)); 
    const currentNote = yield select((store) => (store.notes));
    const remainNote = filterNote(currentNote, itemId);
    yield call(setItemToStorage, 'storageNote', remainNote);
  } catch (error) {
    Alert.alert(
      'Delete Failed',
      String(error),
      null,
      {
        cancelable: false
      }
    );
  }
  yield put({
    type: actions.HIDE_LOADER
  });
}
function* notes () {
//   yield takeEvery('FETCH_NOTES', fetchHandler);
// console.log('This is call from FETCH_NOTES');
  yield take(actions.FETCH_NOTES);
  yield fetchHandler();

  yield takeLatest(actions.addNoteRequest, addNoteHandler);
  yield takeLatest(actions.deleteNoteRequest, deleteNoteHandler);
}
export default function* sagas () {
  yield fork(notes);
  // console.log('index saga');
}