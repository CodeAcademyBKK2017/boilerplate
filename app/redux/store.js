import rootReducer from './reducers/root.reducer';
// import someReduxMiddleware from 'some-redux-middleware';
// import someOtherReduxMiddleware from 'some-other-redux-middleware';
import {applyMiddleware, compose, createStore} from 'redux';
import * as actions from './actions/index.actions';

const enhancerList = [];
const devToolsExtension = window && window.__REDUX_DEVTOOLS_EXTENSION__;

if (typeof devToolsExtension === 'function') {
  enhancerList.push(devToolsExtension());
}

const logger = () => (next) => (action) => {
  // console.log('action is', action);
  next(action);
};

const notAllowToSaveEmptyNote = () => (next) => (action) => {
  if (action.type === actions.ADD_NOTE) {
    if (!action.payload.title || !action.payload.content) {
      // console.log('please save valid note');
    } else {
      next(action);
    }
  } else {
    next(action);
  }
};

const composedEnhancer = compose(applyMiddleware(logger, notAllowToSaveEmptyNote), ...enhancerList);

export const initStore = () => createStore(rootReducer, {}, composedEnhancer);
