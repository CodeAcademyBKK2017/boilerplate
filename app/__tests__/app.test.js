import ApiNotes from '../api';
import ConnectedApp, {mapDispatchToProps}  from '../app';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {Alert} from 'react-native';
import {createStore} from 'redux';
import {getItemToStorage, setItemToStorage} from '../utils/storageutil';
import {shallow} from 'enzyme';

const store = createStore(() => ({}));

jest.mock('Alert', () => ({
  alert: jest.fn()
}));
jest.mock('../api');
jest.mock('../utils/storageutil');

describe('App', () => {
  let wrapper, instance, appComp;
  beforeEach(() => {
    appComp = <ConnectedApp store={store}/>;
    wrapper = shallow(appComp).find('App').shallow();
    instance = wrapper.instance();
    ApiNotes.addNote.mockClear();
    ApiNotes.deleteNote.mockClear();
    ApiNotes.getNotes.mockClear();
    setItemToStorage.mockClear();
    getItemToStorage.mockClear();
  });
  it('renders correctly', () => {
    const tree = renderer.create(appComp);
    expect(tree).toMatchSnapshot();
  });
  it('changeTitle: Should have state Change', () => {
    instance.changeTitle('123');
    expect(instance.state.title).toEqual('123');
  });
  it('changeContent: Should have state Change', () => {
    instance.changeContent('123');
    expect(instance.state.content).toEqual('123');
  });
  it('goToAbout', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.goToAbout();
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenLastCalledWith({'routeName': 'About', 'type': 'Navigation/NAVIGATE'});
  });
  it('onSave success ', async () => {
    const note = {
      title: 'this is title',
      content: 'this is content'
    };
    const expectedNote = {
      title: 'this is title',
      content: 'this is content'
    };
    const expectedNoteWithId = {
      title: 'this is title',
      content: 'this is content',
      id: 1
    };
    const noteList = [expectedNoteWithId];

    instance.setState(note);
    ApiNotes.addNote.mockImplementation(() => Promise.resolve({...note, id: 1}));
    const addNote = jest.fn();
    wrapper.setProps({addNote: addNote});
    await instance.onSave();
    expect(ApiNotes.addNote).toHaveBeenLastCalledWith(expectedNote);
    expect(instance.props.addNote).toHaveBeenLastCalledWith(expectedNoteWithId);
    expect(setItemToStorage).toHaveBeenLastCalledWith('storageNote', noteList);
  });

  it('onSave failure', async () => {
    ApiNotes.addNote.mockImplementation(() => Promise.reject('API failed'));
    await instance.onSave();
    expect(setItemToStorage).not.toBeCalled();
    expect(Alert.alert).toHaveBeenLastCalledWith(
      'Save Failed',
      'API failed',
      [{text: 'OK'}],
      {cancelable: false});
  });
  it('onDelete success', async () => {
    const item = {
      id: 'someid',
      title: 'title x',
      content: 'content x'
    };
    const initialNote = [
      {id: 1, title: 'title y', content: 'content y'},
      item
    ];
    const remainNote = [{id: 1, title: 'title y', content: 'content y'}];
    wrapper.setProps({noteList: initialNote});
    const deleteNote = jest.fn();
    wrapper.setProps({deleteNote: deleteNote});
    const curryFn = instance.onDelete(item);
    await curryFn();
    expect(ApiNotes.deleteNote).toHaveBeenLastCalledWith(item.id);
    expect(setItemToStorage).toHaveBeenLastCalledWith('storageNote', remainNote);
  });
  it('onDelete failure', async () => {
    const item = {
      id: 'failid',
      title: 'fail title',
      content: 'fail message'
    };
    ApiNotes.deleteNote.mockImplementation(() => Promise.reject('API failed'));
    const curryFn = instance.onDelete(item);
    await curryFn();
    expect(Alert.alert).toHaveBeenLastCalledWith(
      'Delete Failed',
      'API failed',
      null,
      {cancelable: false});
  });
  it('init Success Case', async () => {
    const noteList = [{id: 1, title: 'get title', content: 'get content'}];
    ApiNotes.getNotes.mockImplementation(() => Promise.resolve(noteList));
    const populateNote = jest.fn();
    wrapper.setProps({populateNote: populateNote});
    await instance.init();
    expect(ApiNotes.getNotes).toHaveBeenCalled();
    expect(instance.props.populateNote).toHaveBeenLastCalledWith(noteList);
  });
  it('init Fail Case', async () => {
    const note = [{id: 1, title: 'get title', content: 'get content'}];
    ApiNotes.getNotes.mockImplementation(() => Promise.reject('API failed'));
    getItemToStorage.mockImplementation(() => note);
    const populateNote = jest.fn();
    wrapper.setProps({populateNote: populateNote});
    await instance.init();
    expect(getItemToStorage).toHaveBeenLastCalledWith('storageNote');
    expect(instance.props.populateNote).toHaveBeenLastCalledWith(note);
  });
});
