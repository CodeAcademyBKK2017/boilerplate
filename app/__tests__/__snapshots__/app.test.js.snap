// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<View
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "flex-start",
        "padding": 10,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
          "paddingVertical": 10,
          "textAlign": "left",
        }
      }
    >
      Note Title
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderWidth": 1,
          "padding": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 15,
            "fontStyle": "italic",
            "paddingHorizontal": 10,
            "paddingVertical": 10,
            "textAlign": "left",
          }
        }
      >
        Please type your note below
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "fontcustom",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        ï„¨
      </Text>
    </View>
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      style={
        Object {
          "borderWidth": 1,
          "flex": 1,
          "padding": 10,
          "textAlignVertical": "top",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingVertical": 10,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 15,
            "fontWeight": "bold",
          }
        }
      >
        Save
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 15,
        }
      }
    >
      0
       characters
    </Text>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
    >
      about us
    </Text>
  </View>
  <Modal
    animationOutType="fadeOut"
    animationType="none"
    hardwareAccelerated={false}
    onClose={[Function]}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(228, 225, 225, 0.75)",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 0,
          "padding": 20,
        }
      }
      testID={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "padding": 20,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
        testID={undefined}
      >
        <ActivityIndicator
          animating={true}
          color="#2eff00"
          hidesWhenStopped={true}
          size="large"
        />
      </View>
    </View>
  </Modal>
</View>
`;
